### DISCUSSION TITLE
DDR AI characters avoiding battles

### DISCUSSION THREAD

**Discussion by FrozenEmpire** [ChilliHugger/The-Lords-Of-Midnight#216] [Code Algorithm Questions] [answered]
Hi Chris,

I see you made a gameplay rule for AI characters deciding to stay and fight if foes are present...

However is the original DDR behaviour really as follows:
If a player character Attacks another character, then they are preparing for battle and night falls.
Upon night, the AI character who got attacked just vanishes because (a) they have orders to do xyz, and (b) appear to be able to do so at dawn ?

Regards,
Andrew

**Comment by IcemarkUK** [comment, marked as answer]
This is where DDR ai is a little broken, and why the lords just wander around aimlessly sometimes.

All lords are processed for their night before the battles. So reset to dawn and decide their orders etc. They will only fight if they end their turn at a location with an enemy in it.

Their orders are pretty clear -as [worked out here](https://github.com/ChilliHugger/The-Lords-Of-Midnight/blob/35162d67d04e7eaa436aefc9c93467c56cc52bdb/main/midnight/src/tme/scenarios/ddr/ddr_character.cpp#L1087C21-L1087C41). 

But when the lord [picks new orders](https://github.com/ChilliHugger/The-Lords-Of-Midnight/blob/35162d67d04e7eaa436aefc9c93467c56cc52bdb/main/midnight/src/tme/scenarios/ddr/ddr_character.cpp#L1074C1-L1085C2) then there is a small chance of OD_NONE, that's the only chance that they won't move on a turn - other than being at their target location. 

When they decide which direction to travel they then check to see if they can move. If they can't move it's because of one of two reasons, either the location in front of them is blocked by Icy Wastes ( or also technically they are in a tunnel and looking at a wall - but this doesn't actually trigger for AI lords), or the location they are in has enemies

This bit of code in WhatsAtLocation 

```
       MOV    AL,BYTE PTR NoOfEnemysHere
       CMP    AL,000h
       JNZ    @F
       RET
@@:
       XOR    AL,AL
       MOV    BYTE PTR CanWeMoveForward,AL
       RET
```

So the lord then considers why they can't move forward, and if it turns out that it's because they are in a location with an enemy, AND it's DAWN (so their first turn) - it becomes a 50% random check do decide if they stay or if they go. So essentially, if you attack a lord and press night, there is a 50% chance that they just walk away an don't battle occurs.

```
NextMove:                                                      ;L8202
       CALL   TakeObject
       CALL   GetCharLocation
       MOV    WORD PTR WorkingLocation,CX                      ;Where we are
       MOV    DX,WORD PTR ANewLocationToMoveTo                 ;Where we want to go.
      
=== snip working out delta-x, delta-y for target location ===

       MOV    BX,OFFSET L8292-1                                ;Work out best direction
       CALL   A_FromTable                                      ;to look to go where we
       CALL   ChangeLookDirection                              ;want to go.

       CALL   WhatsAtTheLocation
       MOV    AL,BYTE PTR CanWeMoveForward                     ;Are we able to move forward
       CMP    AL,000h
       JNZ    L8257                                            ;yes - so do it
       CALL   GetCharTimeOfDay                                 ;no
       CMP    AL,TIME_DAWN                                     ;would this be the first move of the day?
       JNZ    L8268                                            ;if not, then we will end our turn
       MOV    AL,BYTE PTR LocationBlocked                      ;is location actually enterable (icy wastes)
       CMP    AL,000h
       JZ     L8268                                            ;NO - we can't move so end our turn
       CALL   RandomIshNumber                                  ;YES - so enemy here is stopping the movemnt
       BIT    0,AL                                             ;50:50 chance
       JZ     L8268                                            ;that we end the turn and stop here.
L8257: CALL   MoveCharForward                                  ;walk forward in to location
       CALL   GetCharTimeOfDay
       CMP    AL,TIME_NIGHT                                    ;if it's night, then we need to
       JZ     L8268                                            ;end our turn
       CALL   GetCharTired                                     ;if we still have some energy
       CMP    AL,060h                                          ;then we can take another turn
       JNC    NextMove
L8268: MOV    AL,BYTE PTR NoOfEnemysHere                       ;End TURN
       CMP    AL,001h                                          ;if there is a single enemy here
       JZ     @F                                               ;then we should thinkg about recruiting them
       RET
```

**Reply by FrozenEmpire** [reply]
Thanks for confirming.

50/50 escaping at Dawn seems ok - kind of like the way a player character can decide to escape armies at Dawn too.

The bit I struggle with is when an AI character is first attacked, they have 50/50 chance of disappearing. I think if they are engaged in battle like that, they should be compelled to fight at least 1 night.

(and then more subtle conditions for advanced rules as you have modded)

**Reply by IcemarkUK** [reply]
The difference is for player character escaping at dawn is AFTER a fight, for AI it's before. So when you attack there is only a 50% chance to will actually fight the lord you are attacking. I think the 50:50 check would be fine for a coward, I also think it would be fine for an attack that came from an approach. ie: It wasn't a direct assault  so why attack back... but if the enemy is there because of an attack then maybe the 50:50 check is harsh.

I think this is a gameplay bug that should have been caught in play testing, but I doubt there was enough time. I also suspect Mike didn't have the observability on what the game was doing, like we might have now. I really miss Mike when talking about these kind of nuances. I'd love to know what he though he was thinking!

For me I've already pushed a gameplay rule to adjust this. But I think I'm going to tweak it some more.

**Reply by FrozenEmpire** [reply]
Good point on the observability we have these days compared to back then, easy to forget that !

I'm tempted to introduce a new battle flag "HasBeenAttacked" which gets set on all foes when a player character "Attacks" ie enters battle. And then use this to hold the AI characters in position for the 1st night. The next dawn, they can decide to flee if they wish :)

I guess this is one key difference to LOM. DDR doesn't have that "lets check out something interesting" logic that makes AI characters check out other characters around them. Now that would be a good advanced gameplay rule ;)

**Reply by IcemarkUK** [reply]
Yes, I am thinking of the same. But I'm also considering looking at the order in which battles are triggered - not from top to bottom, but attackers first. So players lords that attack and AI lords that enter a location to attack. Currently it's possible for a defender to get first initiative. And this matters when it's lord on lord, because it can be so easy for the attacking lord to kill the defending lord.

Also an HasBeenApproached which changes the condition. So a failed approach will result in an attack, but an approach gives the approached lord chance to move away rather than continue with the attack. If a lord approaches and fails the initiative for the attack should there be one should be on the approached lord.

> DDR doesn't have that "lets check out something interesting" logic that makes AI characters check out other characters around them. Now that would be a good advanced gameplay rule 

I added that over the weekend, but tweaked it for DDR in that AI lords are more prone to be interested in the players lords rather than the other AI lords. not pushed it yet. I was thinking of adding a slight tweak so that the race of their enemy is considered so you'd have something like 75% check out player's lords, 50% check out foe's lords, and 25% everyone else.
